core/distributed/matrix.cpp

```diff
@@ -0,0 +1,440 @@
+    result->set_size(this->get_size());
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Matrix<ValueType, LocalIndexType, GlobalIndexType>::move_to(
+    Matrix<next_precision<value_type>, local_index_type, global_index_type>*
+        result)
+{
+    GKO_ASSERT(this->get_communicator().size() ==
+               result->get_communicator().size());
+    result->local_mtx_->move_from(this->local_mtx_.get());
+    result->non_local_mtx_->move_from(this->non_local_mtx_.get());
+    result->gather_idxs_ = std::move(this->gather_idxs_);
+    result->send_offsets_ = std::move(this->send_offsets_);
+    result->recv_offsets_ = std::move(this->recv_offsets_);
+    result->recv_sizes_ = std::move(this->recv_sizes_);
+    result->send_sizes_ = std::move(this->send_sizes_);
+    result->non_local_to_global_ = std::move(this->non_local_to_global_);
+    result->set_size(this->get_size());
+    this->set_size({});
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
+    const device_matrix_data<value_type, global_index_type>& data,
+    const Partition<local_index_type, global_index_type>* row_partition,
+    const Partition<local_index_type, global_index_type>* col_partition)
+{
+    const auto comm = this->get_communicator();
+    GKO_ASSERT_EQ(data.get_size()[0], row_partition->get_size());
+    GKO_ASSERT_EQ(data.get_size()[1], col_partition->get_size());
+    GKO_ASSERT_EQ(comm.size(), row_partition->get_num_parts());
+    GKO_ASSERT_EQ(comm.size(), col_partition->get_num_parts());
+    auto exec = this->get_executor();
+    auto local_part = comm.rank();
+
+    // set up LinOp sizes
+    auto num_parts = static_cast<size_type>(row_partition->get_num_parts());
+    auto global_num_rows = row_partition->get_size();
+    auto global_num_cols = col_partition->get_size();
+    dim<2> global_dim{global_num_rows, global_num_cols};
+    this->set_size(global_dim);
+
+    // temporary storage for the output
+    array<local_index_type> diag_row_idxs{exec};
+    array<local_index_type> diag_col_idxs{exec};
+    array<value_type> diag_values{exec};
+    array<local_index_type> offdiag_row_idxs{exec};
+    array<local_index_type> offdiag_col_idxs{exec};
+    array<value_type> offdiag_values{exec};
+    array<local_index_type> recv_gather_idxs{exec};
+    array<comm_index_type> recv_offsets_array{exec, num_parts + 1};
+
+    // build diagonal, off-diagonal matrix and communication structures
+    exec->run(matrix::make_build_diag_offdiag(
+        data, make_temporary_clone(exec, row_partition).get(),
+        make_temporary_clone(exec, col_partition).get(), local_part,
+        diag_row_idxs, diag_col_idxs, diag_values, offdiag_row_idxs,
+        offdiag_col_idxs, offdiag_values, recv_gather_idxs,
+        recv_offsets_array.get_data(), non_local_to_global_));
```

Comment:
I think it is better to be uniform here with either all parameters as pointers or all as `array`s

Reply:
for recv_offsets it's not necessary to use array, because the array size is known beforehand, but I will still switch it to array.

Reply:
I've been somewhat consistently using the rule: If it needs to be resized, use array&, otherwise use pointer. Not sure if that makes sense, but I'll just mention it here